Introduce conductor to orchestrate our nightly pipelines & deployments

chouquette opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?

This PR prepares for conductor and adds a service definition file for that purpose.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://datadoghq.atlassian.net/browse/APL-2694

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

chouquette committed:
CI: forward DDR_WORKFLOW_ID variable to downstream pipelines

chouquette committed:
add an initial service definition file

spencergilbert commented on service.datadog.yaml:
```diff
@@ -0,0 +1,13 @@
+      schedule: "0 3 3 * * 1-5"
+      env:
+        - name: "Nightly builds"
+          branch: "chouquette/conductor"
```
Presumably this is intentional for now and it should be `main` once the feature is stabilized?

chouquette replied:
Indeed maybe it makes more sense to merge without this, and add it once the service is known to DDR

sblumenthal commented on service.datadog.yaml:
```diff
@@ -0,0 +1,13 @@
+---
```
This file will conflict with the service definition for the agent defined [here](https://github.com/DataDog/k8s-datadog-agent-ops/blob/main/service.datadog.yaml#L3), and it is used to power the datadog-agent entry in the service catalog

I don't know if the behavior is that they will keep replacing each other, or if both will be merged in the backend, but just in case, is it possible to rename the dd-service to something like `datadog-agent-test` for now?

chouquette replied:
renamed to `datadog-agent-nightly`

sblumenthal commented on .gitlab/install_script_testing.yml:
```diff
@@ -10,7 +10,7 @@ test_install_script:
     - export TESTING_YUM_URL=$RPM_TESTING_S3_BUCKET
     - export TEST_PIPELINE_ID=$CI_PIPELINE_ID
-    - inv pipeline.trigger-child-pipeline --project-name "DataDog/agent-linux-install-script" --git-ref "main" --variables "TESTING_APT_URL,TESTING_YUM_URL,TEST_PIPELINE_ID"
+    - inv pipeline.trigger-child-pipeline --project-name "DataDog/agent-linux-install-script" --git-ref "main" --variables "TESTING_APT_URL,TESTING_YUM_URL,TEST_PIPELINE_ID,DDR_WORKFLOW_ID"
```
This does not JUST apply to this file, it applies to every change where you have updated the `trigger-child-pipeline` call: My understanding is that this will error out because the invoke task does not like undefined variables, and this variable will be undefined when not run by conductor

See https://github.com/DataDog/datadog-agent/pull/21690

chouquette replied:
Wouldn't it be easier to adjust the invoke task to not attempt to forward unknown environment variables?

chouquette replied:
Ah I missed your note in the linked PR, and you have a good point. Maybe an additional parameter to conditionally forward variables if they exist could be a good compromise? Or emiting a warning frem the invoke task when an unknown variable is requested for forwarding?

chouquette replied:
I went the warning way for now but I'm open for alternative approches

sblumenthal replied:
I am happy with changing the behavior, I just wasn't sure if it was intentional or not. I guess no jobs currently fail to startup in this manner, so it shouldn't break the existing pipeline, but for people who are adding a new job it could be a pain to debug

Which could be solved with a warning, so that works!

sblumenthal commented on .gitlab/deploy_containers.yml:
```diff
@@ -24,17 +25,19 @@ deploy_containers-a6-on-failure:
     PARENT_PIPELINE_ID: $CI_PIPELINE_ID
     BUCKET_BRANCH: $BUCKET_BRANCH
     FORCE_MANUAL: "true"
+    DDR_WORKFLOW_ID: $DDR_WORKFLOW_ID
```
Unless this changed with the new version of gitlab, redefining the same variable name as an already defined variable caused the pipeline to error out when run

chouquette replied:
I'll try to see if this was reported/fixed, but that's indeed the case then I guess we might want to systematically use the invoke task to trigger downstream pipelines

chouquette replied:
It turns out the triggered pipeline already rely on the iknvoke task so there's no need to forward it explicitely from here

sblumenthal commented on .gitlab/docker_common/publish_job_templates.yml:
```diff
@@ -18,4 +18,4 @@
     - IMG_VARIABLES="$(sed -E "s#(${SRC_AGENT}|${SRC_DSD}|${SRC_DCA}|${SRC_CWS_INSTRUMENTATION})#\1${ECR_RELEASE_SUFFIX}#g" <<<"$IMG_VARIABLES")"
     - IMG_SOURCES="$(sed -E "s#(${SRC_AGENT}|${SRC_DSD}|${SRC_DCA}|${SRC_CWS_INSTRUMENTATION})#\1${ECR_RELEASE_SUFFIX}#g" <<<"$IMG_SOURCES")"
-    - inv pipeline.trigger-child-pipeline --project-name "DataDog/public-images" --git-ref "main" --variables "IMG_VARIABLES,IMG_REGISTRIES,IMG_SOURCES,IMG_DESTINATIONS,IMG_SIGNING"
+    - inv pipeline.trigger-child-pipeline --project-name "DataDog/public-images" --git-ref "main" --variables "IMG_VARIABLES,IMG_REGISTRIES,IMG_SOURCES,IMG_DESTINATIONS,IMG_SIGNING,DDR_WORKFLOW_ID"
```
You are going to want to pass the full list of conductor-defined env vars to the downstream pipeline:
```
APPS,BAZEL_TARGET,DDR,DDR_WORKFLOW_ID,TARGET_ENV,DYNAMIC_BUILD_RENDER_TARGET,DYNAMIC_BUILD_RENDER_TARGET_FORWARD_PARAMETERS
```

sblumenthal reviewed (commented):
I added a bunch of comments based on issues I ran into when trying to set this up

chouquette committed:
tasks: pipeline: don't fail when forwarding undefined variable

chouquette committed:
rename service to avoid conflicting with the existing one

chouquette committed:
forward all conductor variables to downstream pipelines

chouquette committed:
remove unneeded variables forwarding

all conductor variables are already being forwarded by
.docker_publish_job_definition

chouquette committed:
fix yaml syntax

chouquette committed:
don't attempt to use an empty env variable

chouquette committed:
more yaml fixes
